File: components/gmaps-flow/version-1.0/google-map-flow/src/main/java/com/neotropic/flow/component/googlemap/OverlayView.java
/*
 *  Copyright 2010-2020 Neotropic SAS <contact@neotropic.co>.
 *  
 *  Licensed under the Apache License, Version 2.0 (the "License"); 
 *  you may not use this file except in compliance with the License. 
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software 
 *  distributed under the License is distributed on an "AS IS" BASIS, 
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
 *  See the License for the specific language governing permissions and 
 *  limitations under the License.
 */
package com.neotropic.flow.component.googlemap;

import com.vaadin.flow.component.Component;
import com.vaadin.flow.component.ComponentEventListener;
import com.vaadin.flow.component.HasComponents;
import com.vaadin.flow.component.Tag;
import com.vaadin.flow.component.dependency.JsModule;
import com.vaadin.flow.shared.Registration;
import elemental.json.Json;
import elemental.json.JsonArray;
import elemental.json.JsonObject;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Objects;
import java.util.function.Consumer;

/**
 *
 * @author Johny Andres Ortega Ruiz {@literal <johny.ortega@kuwaiba.org>}
 */
@Tag("overlay-view")
@JsModule("./overlay-view.js")
public class OverlayView extends Component implements HasComponents {
        
    public OverlayView() {
    }
    
    public OverlayView(boolean mapBounds) {
        getElement().setProperty(Constants.Property.MAP_BOUNDS, mapBounds);
    }
    
    public OverlayView(LatLngBounds bounds) {
        if (bounds != null)
            getElement().setPropertyJson(Constants.Property.BOUNDS, bounds.toJson());
    }
    
    public LatLngBounds getBounds() {
        return new LatLngBounds(
            (JsonObject) getElement().getPropertyRaw(Constants.Property.BOUNDS)
        );
    }
    
    public void setBounds(LatLngBounds bounds) {
        if (bounds != null)
            getElement().setPropertyJson(Constants.Property.BOUNDS, bounds.toJson());
        else
            getElement().setPropertyJson(Constants.Property.BOUNDS, Json.createNull());
    }
    
    public boolean getMapBounds() {
        return getElement().getProperty(Constants.Property.MAP_BOUNDS, false);
    }
    
    public void setMapBounds(boolean mapBounds) {
        getElement().setProperty(Constants.Property.MAP_BOUNDS, mapBounds);
    }
    
    public void fromLatLngToDivPixel(LatLng latLng, Consumer<Point> consumer) {
        getElement()
            .executeJs("return this.fromLatLngToDivPixel($0, $1)", latLng.getLat(), latLng.getLng())
            .then(JsonObject.class, point -> {
                consumer.accept(new Point(
                    point.getNumber(Constants.Property.X), 
                    point.getNumber(Constants.Property.Y)
                ));
            });
    }
    
    public void fromDivPixelToLatLng(Point point, Consumer<LatLng> consumer) {
        getElement()
            .executeJs("return this.fromDivPixelToLatLng($0, $1)", point.getX(), point.getY())
            .then(JsonObject.class, latLng -> {
                consumer.accept(new LatLng(
                    latLng.getNumber(Constants.Property.LAT),
                    latLng.getNumber(Constants.Property.LNG)
                ));
            });
    }
    /**
     * Executes callback to get the pixel coordinates of the given geographical location.
     * @param geoCoordinates List of geographical coordinates.
     * @param callback Callback to get the pixel coordinates of the given geographical location.
     */
    public void fromLatLngToDivPixel(List<LatLng> geoCoordinates, Consumer<List<Point>> callback) {
        Objects.requireNonNull(geoCoordinates);
        Objects.requireNonNull(callback);
        if (!geoCoordinates.isEmpty()) {
            JsonArray array = Json.createArray();
            for (int i = 0; i < geoCoordinates.size(); i++)
                array.set(i, geoCoordinates.get(i).toJson());
            getElement()
                .executeJs("return this.projectionFromLatLngToDivPixel($0)", array) //NOI18N
                .then(JsonArray.class, pixelCoordinates -> {
                    if (pixelCoordinates != null && pixelCoordinates.length() > 0) {
                        List<Point> list = new ArrayList();
                        for (int i = 0; i < pixelCoordinates.length(); i++)
                            list.add(new Point(pixelCoordinates.getObject(i)));
                        callback.accept(list);
                    }
                    else
                        callback.accept(null);
                }
            );
        }
    }
    /**
     * Executes callback to get the pixel coordinates of the given geographical location.
     * @param geoCoordinates List of geographical coordinates.
     * @param callback Callback to get the pixel coordinates of the given geographical location.
     */
    public void fromLatLngToDivPixel(HashMap<String, List<LatLng>> geoCoordinates, Consumer<HashMap<String, List<Point>>> callback) {
        Objects.requireNonNull(geoCoordinates);
        Objects.requireNonNull(callback);
        if (!geoCoordinates.isEmpty()) {
            String[] ids = geoCoordinates.keySet().toArray(new String[0]);
            /**
             * JSON Object
             * {
             *   "id": [{"lat": 0.0, "lng": 0.0}, ..],
             *   ..
             * }
             */
            JsonObject object = Json.createObject();
            for (int i = 0; i < ids.length; i++) {
                List<LatLng> coordinates = geoCoordinates.get(ids[i]);
                if (!coordinates.isEmpty()) {
                    JsonArray jsonCoordinates = Json.createArray();
                    for (int j = 0; j < coordinates.size(); j++) {
                        LatLng coordinate = coordinates.get(j);
                        JsonObject jsonCoordinate = Json.createObject();
                        jsonCoordinate.put(Constants.JsonKey.LAT, coordinate.getLat());
                        jsonCoordinate.put(Constants.JsonKey.LNG, coordinate.getLng());
                        jsonCoordinates.set(j, jsonCoordinate);
                    }
                    object.put(ids[i], jsonCoordinates);
                }
            }
            if (object.keys().length > 0) {
                getElement()
                    .executeJs("return this.projectionMapFromLatLngToDivPixel($0)", object)
                    .then(JsonObject.class, pixelCoordinates -> {
                        /**
                         * pixelCoordinates
                         * {
                         *   "id": [{"x": 0.0, "y": 0.0}, ..],
                         *   ..
                         * }
                         */
                        if (pixelCoordinates != null && pixelCoordinates.keys().length > 0) {
                            HashMap<String, List<Point>> points = new HashMap();
                            for (int i = 0; i < pixelCoordinates.keys().length; i++) {
                                String id = pixelCoordinates.keys()[i];
                                JsonArray jsonPoints = pixelCoordinates.getArray(id);
                                List<Point> idPoints = new ArrayList();
                                for (int j = 0; j < jsonPoints.length(); j++) {
                                    JsonObject jsonPoint = jsonPoints.getObject(j);
                                    idPoints.add(new Point(
                                        jsonPoint.getNumber(Constants.JsonKey.X),
                                        jsonPoint.getNumber(Constants.JsonKey.Y)
                                    ));
                                }
                                points.put(id, idPoints);
                            }
                            callback.accept(points);
                        }
                        else
                            callback.accept(null);
                    }
                );
            }
        }
    }
    /**
     * Executes callback to get the geographical coordinates from pixel coordinates.
     * @param pixelCoordinates List of pixel coordinates.
     * @param callback Callback to get the geographical coordinates from pixel coordinates.
     */
    public void fromDivPixelToLatLng(List<Point> pixelCoordinates, Consumer<List<LatLng>> callback) {
        Objects.requireNonNull(pixelCoordinates);
        Objects.requireNonNull(callback);
        if (!pixelCoordinates.isEmpty()) {
            JsonArray array = Json.createArray();
            for (int i = 0; i < pixelCoordinates.size(); i++)
                array.set(i, pixelCoordinates.get(i).toJson());
            getElement()
                .executeJs("return this.projectionFromDivPixelToLatLng($0)", array) //NOI18N
                .then(JsonArray.class, geoCoordinates -> {
                    if (geoCoordinates != null && geoCoordinates.length() > 0) {
                        List<LatLng> list = new ArrayList();
                        for (int i = 0; i < geoCoordinates.length(); i++)
                            list.add(new LatLng(geoCoordinates.getObject(i)));
                        callback.accept(list);
                    }
                    else
                        callback.accept(null);
                }
            );
        }
    }
    /**
     * Executes callback to get the geographical coordinates from pixel coordinates.
     * @param pixelCoordinates List of pixel coordinates.
     * @param callback Callback to get the geographical coordinates from pixel coordinates.
     */
    public void fromDivPixelToLatLng(HashMap<String, List<Point>> pixelCoordinates, Consumer<HashMap<String, List<LatLng>>> callback) {
        Objects.requireNonNull(pixelCoordinates);
        Objects.requireNonNull(callback);
        if (!pixelCoordinates.isEmpty()) {
            String[] ids = pixelCoordinates.keySet().toArray(new String[0]);
            /**
             * JSON Object
             * {
             *   "id": [{"x": 0.0, "y": 0.0}, ..],
             *   ..
             * }
             */
            JsonObject object = Json.createObject();
            for (int i = 0; i < ids.length; i++) {
                List<Point> points = pixelCoordinates.get(ids[i]);
                if (!points.isEmpty()) {
                    JsonArray jsonPoints = Json.createArray();
                    for (int j = 0; j < points.size(); j++) {
                        Point point = points.get(j);
                        JsonObject jsonPoint = Json.createObject();
                        jsonPoint.put(Constants.JsonKey.X, point.getX());
                        jsonPoint.put(Constants.JsonKey.Y, point.getY());
                        jsonPoints.set(j, jsonPoint);
                    }
                    object.put(ids[i], jsonPoints);
                }
            }
            if (object.keys().length > 0) {
                getElement()
                    .executeJs("return this.projectionMapFromDivPixelToLatLng($0)", object)
                    .then(JsonObject.class, geoCoordinates -> {
                        /**
                         * geoCoordinates
                         * {
                         *   "id": [{"lat": 0.0, "lng": 0.0}, ..]
                         *   ..
                         * }
                         */
                        if (geoCoordinates != null && geoCoordinates.keys().length > 0) {
                            HashMap<String, List<LatLng>> coordinates = new HashMap();
                            for (int i = 0; i < geoCoordinates.keys().length; i++) {
                                String id = geoCoordinates.keys()[i];
                                JsonArray jsonCoordinates = geoCoordinates.getArray(id);
                                List<LatLng> idCoordinates = new ArrayList();
                                for (int j = 0; j < jsonCoordinates.length(); j++) {
                                    JsonObject jsonCoordinate = jsonCoordinates.getObject(j);
                                    idCoordinates.add(new LatLng(
                                        jsonCoordinate.getNumber(Constants.JsonKey.LAT),
                                        jsonCoordinate.getNumber(Constants.JsonKey.LNG)
                                    ));
                                }
                                coordinates.put(id, idCoordinates);
                            }
                            callback.accept(coordinates);
                        }
                        else
                            callback.accept(null);
                    }
                );
            }
        }
    }
    public Registration addWidthChangedListener(ComponentEventListener<GoogleMapEvent.OverlayViewWidthChangedEvent> listener) {
        return addListener(GoogleMapEvent.OverlayViewWidthChangedEvent.class, listener);
    }
}

File: components/gmaps-flow/branches/version-2.5/gmaps-flow-root/gmaps-flow/src/main/resources/META-INF/resources/frontend/overlay-view.js
/** 
@license
Copyright 2010-2019 Neotropic SAS <contact@neotropic.co>.

Licensed under the Apache License, Version 2.0 (the "License"); 
you may not use this file except in compliance with the License. 
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software 
distributed under the License is distributed on an "AS IS" BASIS, 
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. 
See the License for the specific language governing permissions and 
limitations under the License.
*/
import {html, PolymerElement} from '@polymer/polymer/polymer-element.js';
import {FlattenedNodesObserver} from '@polymer/polymer/lib/utils/flattened-nodes-observer.js'

/**
 * `<overlay-view>`
 * @customElement
 * @polymer
 * @author Johny Andres Ortega Ruiz {@literal <johny.ortega@kuwaiba.org>}
 */
class OverlayView extends PolymerElement {
  static get is() {
    return 'overlay-view';
  }

  static get template() {
    return html`
      <style>
        :host {
          display: block;
        }
      </style>
    `;
  }

  static get properties() {
    return {
      /**
       * Rectangle bounds
       * @type {{east: number, north: number, south: number, west: number}}
       */
      bounds: {
        type: Object,
        observer: '_boundsChanged'
      },
      /**
       * Bounds of the current viewport.
       */
      mapBounds: {
        type: Boolean,
        value: false
      }
    };
  }
  /**
   * @return {google.maps.OverlayView} The google.maps.OverlayView
   */
  getMVCObject() {
    return this.overlay;
  }
  /**
   * @param {google.maps.Map} map 
   */
  added(map) {
    MyOverlay.prototype = new google.maps.OverlayView();

    function MyOverlay(bounds, map) {
      this._bounds = bounds;
      this._div = null;
      this.setMap(map);
    }
    var _this = this;

    MyOverlay.prototype.onAdd = function() {
      this._div = _this._div;
      this.getPanes().overlayLayer.appendChild(this._div);
    }

    MyOverlay.prototype.draw = function() {
      _this.projection = this.getProjection();
      if (_this.mapBounds) {
        var sw = _this.projection.fromLatLngToDivPixel(this.getMap().getBounds().getSouthWest());
        var ne = _this.projection.fromLatLngToDivPixel(this.getMap().getBounds().getNorthEast());
      } else {
        var sw = _this.projection.fromLatLngToDivPixel(this._bounds.getSouthWest());
        var ne = _this.projection.fromLatLngToDivPixel(this._bounds.getNorthEast());
      }
      var div = this._div;
      div.style.position = 'absolute';
      div.style.left = sw.x + 'px';
      div.style.top = ne.y + 'px';
      var width = ne.x - sw.x;
      var height = sw.y - ne.y;
      div.style.width = width + 'px';
      div.style.height = height + 'px';    

      _this.dispatchEvent(new CustomEvent('overlay-view-width-changed', 
        {
          detail: {
            width: width
          }
        }
      ));         
    }
    
    MyOverlay.prototype.onRemove = function() {
      this._div.parentNode.removeChild(this._div);
      this._div = null;
    }

    this._observer = new FlattenedNodesObserver(this, info => {
      info.addedNodes.forEach(value => {
        this._div = value;
        if (this.mapBounds) {
          this.overlay = new MyOverlay(null, map);
        } else {
          this.overlay = new MyOverlay(
            new google.maps.LatLngBounds(
              {lat: this.bounds.south, lng: this.bounds.west}, 
              {lat: this.bounds.north, lng: this.bounds.east}
            ), 
            map
          );
        }
      });
      info.removedNodes.forEach(value => {
      });
    });
  }
  /**
   * Gets the pixel coordinates of the given geographical location.
   * @param {number} lat - Latitude.
   * @param {number} lng - Longitude.
   * @returns {Object} Pixel coordinates {x: 0.0, y: 0.0}.
   */
  fromLatLngToDivPixel(lat, lng) {
    if (this.projection) {
      var pixel = this.projection.fromLatLngToDivPixel(new google.maps.LatLng({lat: lat, lng: lng}));
      return {x: pixel.x, y: pixel.y};
    }
    return null;
  }
  /**
   * Gets the geographical coordinates from pixel coordinates.
   * @param {number} x - Pixel coordinate x.
   * @param {number} y - Pixel coordinate y.
   * @returns {Object} Geographical coordinates {lat: 0.0, lng: 0.0}.
   */
  fromDivPixelToLatLng(x, y) {
    if (this.projection) {
      var latLng = this.projection.fromDivPixelToLatLng(new google.maps.Point(x, y));
      return {lat: latLng.lat(), lng: latLng.lng()}
    }
    return null;
  }
  /**
   * Gets the pixel coordinates of the given geographical location.
   * @param {Array} geoCoordinates - Geographical coordinates Array [{lat: 0.0, lng: 0.0}, ..].
   * @returns {Array} Pixel coordinates Array [{x: 0.0, y: 0.0}, ..].
   */
  projectionFromLatLngToDivPixel(geoCoordinates) {
    if (this.projection && geoCoordinates) {
      var pixelCoordinates = [];
      geoCoordinates.forEach(geoCoordinate => {
        var pixelCoordinate = this.projection.fromLatLngToDivPixel(new google.maps.LatLng({lat: geoCoordinate.lat, lng: geoCoordinate.lng}));
        pixelCoordinates.push({x: pixelCoordinate.x, y: pixelCoordinate.y});
      });
      return pixelCoordinates;
    }
    return null;
  }
  /**
   * Gets the geographical coordinates from pixel coordinates.
   * @param {Array} pixelCoordinates - Pixel Array [{x: 0.0, y: 0.0}, ..].
   * @returns {Array} Geographical coordinates Array [{lat: 0.0, lng: 0.0}, ..].
   */
  projectionFromDivPixelToLatLng(pixelCoordinates) {
    if (this.projection && pixelCoordinates) {
      var geoCoordinates = [];
      pixelCoordinates.forEach(pixelCoordinate => {
        var geoCoordinate = this.projection.fromDivPixelToLatLng(new google.maps.Point(pixelCoordinate.x, pixelCoordinate.y));
        geoCoordinates.push({lat: geoCoordinate.lat(), lng: geoCoordinate.lng()});
      });
      return geoCoordinates;
    }
    return null;
  }
  /**
   * Gets the pixel coordinates of the given geographical location.
   * @param {Array} geoCoordinates - Geographical coordinates Object {"id": [{lat: 0.0, lng: 0.0}, ..] ..}.
   * @returns {Array} Pixel coordinates Object {"id": [{x: 0.0, y: 0.0}, ..] ..}.
   */
  projectionMapFromLatLngToDivPixel(geoCoordinates) {
    if (this.projection && geoCoordinates) {
      var pixelCoordinates = {};
      for (const key in geoCoordinates) {
        pixelCoordinates[key] = this.projectionFromLatLngToDivPixel(geoCoordinates[key]);
      }
      return pixelCoordinates;
    }
    return null;
  }
  /**
   * Gets the geographical coordinates from pixel coordinates.
   * @param {Array} pixelCoordinates - Pixel coordinates Object {"id": [{x: 0.0, y: 0.0}, ..], ..}.
   * @returns {Array} Geographical coordinates Object {"id": [{lat: 0.0, lng: 0.0}, ..], ..}.
   */
  projectionMapFromDivPixelToLatLng(pixelCoordinates) {
    if (this.projection && pixelCoordinates) {
      var geoCoordinates = {};
      for (const key in pixelCoordinates) {
        geoCoordinates[key] = this.projectionFromDivPixelToLatLng(pixelCoordinates[key]);
      }
      return geoCoordinates;
    }
    return null;
  }

  removed() {
    if (this.overlay)
      this.overlay.setMap(null);
  }

  _boundsChanged(newValue) {
    if (this.overlay) {
      this.overlay._bounds = new google.maps.LatLngBounds(
        {lat: newValue.south, lng: newValue.west},
        {lat: newValue.north, lng: newValue.east}
      );
    }
  }
}

window.customElements.define(OverlayView.is, OverlayView);